Add REST API v2

dimas-b opened the pull request:
Add v2 REST endpoints.

Add java client implementation for v2 REST API.
Since namespaces are treated as ordinary content objects in API v2
the old java client API for dealing with namespaces is implemented
on the client side on top of v2 "get entries" and "commit" APIs.

Run in-memory REST server tests for both v1 and v2 APIs.

REST test using persistent backend only use v2 API.

Resteasy test remain on v1 API since they construct REST requests
according to Nessie API v1 specs.

OpenAPI YAML is built and published only for v2. OpenAPI for v1
should be obtained from older Nessie publications.

This is the main contribution towards #5112

----

This PR builds on top of #5419, #5420, #5435, #5436.

dimas-b commented on ui/build.gradle.kts:
```diff
@@ -34,7 +34,10 @@ plugins {
+// Use  Nessie REST API V1 in UI for now. TODO: Migrate to current Nessie REST API from `model`
+dependencies { openapi(create("org.projectnessie", "nessie-model", "0.40.2").setTransitive(false)) }
+// Resolve the openapi dependency immediately to avoid accidental override during integration tests
+openapi.resolve()
```
@snazy : This PR is not ready for review yet, but I'm wondering if you have any suggestions for this particular build script change - without `resolve` integration test would fail because they pulled the `current` Nessie model jar (e.g. 0.41.1-SNAPSHOT`).

snazy replied:
`.resolve` is a no go. If you're going to change the UI anyway, you can get away with the following for now.
That Download task isn't awesome, because it will perform a HTTP HEAD everytime, but maybe good enough as a _temporary_ workaround.

```
diff --git a/ui/build.gradle.kts b/ui/build.gradle.kts
index 14e610b1f..2874d9795 100644
--- a/ui/build.gradle.kts
+++ b/ui/build.gradle.kts
@@ -19,6 +19,7 @@ import com.github.gradle.node.npm.task.NpmInstallTask
 import com.github.gradle.node.npm.task.NpmSetupTask
 import com.github.gradle.node.npm.task.NpmTask
 import com.github.gradle.node.task.NodeSetupTask
+import fi.linuxbox.gradle.download.Download
 
 plugins {
   // The ':nessie-ui' module is technically not a Java library, but declaring it as such provides
@@ -30,15 +31,9 @@ plugins {
   signing
   id("com.github.node-gradle.node")
   `nessie-conventions`
+  id("fi.linuxbox.download") version "0.6"
 }
 
-val openapi by configurations.creating { description = "Used to reference OpenAPI spec files" }
-
-// Use  Nessie REST API V1 in UI for now. TODO: Migrate to current Nessie REST API from `model`
-dependencies { openapi(create("org.projectnessie", "nessie-model", "0.40.2").setTransitive(false)) }
-// Resolve the openapi dependency immediately to avoid accidental override during integration tests
-openapi.resolve()
-
 node {
   download.set(true)
   version.set("16.14.2")
@@ -67,17 +62,11 @@ val clean =
   }
 
 val pullOpenApiSpec by
-  tasks.registering(Sync::class) {
-    destinationDir = openApiSpecDir
-    from(project.objects.property(Configuration::class).value(openapi))
-
-    doLast {
-      copy {
-        openApiSpecDir.listFiles { f -> f.name.endsWith(".jar") }?.forEach { f -> from(zipTree(f)) }
-        into(openApiSpecDir)
-        include("META-INF/openapi/openapi.yaml")
-      }
-    }
+  tasks.registering(Download::class) {
+    val downloadDir = openApiSpecDir.resolve("META-INF/openapi")
+    val v1Openapi = downloadDir.resolve("openapi.yaml")
+    from("https://github.com/projectnessie/nessie/releases/download/nessie-0.42.0/nessie-openapi-0.42.0.yaml")
+    to(v1Openapi)
   }
 
 val nodeSetup =
```

dimas-b replied:
My thinking was to do UI changes in a follow-up PR. Will switch to the download task for now.

dimas-b replied:
Alternative approach: #5408

dimas-b replied:
implemented differently in #5408

snazy commented on model/src/main/java/org/projectnessie/api/TreeApi.java:
```diff
@@ -129,38 +144,47 @@ LogResponse getCommitLog(
+   * @param params The {@link DiffParams} that includes the parameters for this API call.
+   * @return A set of diff values that show the difference between two given references.
+   */
+  DiffResponse getDiff(@Valid @NotNull DiffParams params) throws NessieNotFoundException;
```
Let's make `DiffResponse` pageable in v2 (just adding the request/response parameters now)

dimas-b replied:
#5423 - I'd like to keep this PR limited to v2 API that cover the same functionality that also exists in v1.

snazy commented on model/src/main/java/org/projectnessie/api/TreeApi.java:
```diff
@@ -192,9 +213,48 @@ Branch commitMultipleOperations(
+   * @return list of {@link GetMultipleContentsResponse.ContentWithKey}s
+   * @throws NessieNotFoundException if {@code ref} or {@code hashOnRef} does not exist
+   */
+  GetMultipleContentsResponse getMultipleContents(
```
Wonder whether we should change the `CommitMeta` for v2 to be more generic, so these attributes:
* commit message
* map of property name to list of property values
* commit ID
* list of parent commit IDs
Pretty much like Git does. With that, there could be for example a list of authors (think: squashing & merging). WDYT?

dimas-b replied:
makes sense :+1:

dimas-b replied:
#5424 - I'd like to keep this PR limited to v2 API that cover the same functionality that also exists in v1.

snazy commented on .idea/codeStyles/Project.xml:
```diff
@@ -1,13 +1,5 @@
 <component name="ProjectCodeStyleConfiguration">
   <code_scheme name="Project" version="173">
-    <option name="OTHER_INDENT_OPTIONS">
```
(Heads up: remove the IDE file changes)

dimas-b replied:
oops - will remove. Actually, I think I'll split into multiple PRs once it's ready.

snazy commented on model/build.gradle.kts:
```diff
@@ -47,16 +47,11 @@ dependencies {
   operationIdStrategy.set("METHOD")
   scanPackages.set(
-    listOf("org.projectnessie.api", "org.projectnessie.api.http", "org.projectnessie.model")
+    listOf("^(org.projectnessie.api(\\..+)?|org.projectnessie.api.http|org.projectnessie.model)$")
```
Should the v1/v2 interfaces/classes reside in `org.projectnessie.api.v1`/`...api.v2` instead of `...apiv1`?

dimas-b replied:
I wanted to keep the "current" api as simply `api` with the idea of dropping `apiv1` once clients migrate to v2.

Do you see us supporting both versions for a long enough time to benefit from parallel `v1` and `v2` packages?

dimas-b replied:
On the other hand, it might be worth splitting java packages by API version anyway, just to make it clear at compile time... will give it a try :thinking:

dimas-b replied:
fixed

XN137 commented on model/src/main/java/org/projectnessie/api/v2/params/ReferencesParams.java:
```diff
@@ -0,0 +1,135 @@
+import javax.ws.rs.QueryParam;
+import org.eclipse.microprofile.openapi.annotations.media.ExampleObject;
+import org.eclipse.microprofile.openapi.annotations.parameters.Parameter;
+import org.projectnessie.api.params.FetchOption;
```
isnt it odd that this class is the only one remaining below `api` ?
will we keep it there for eternity i.e. even if one day we can remove `api.v1` ?

dimas-b replied:
It is odd - I commented in the commit about that. It looks like a mistake to me that it was originally added under `param`. It should have been under `model`, IMHO. However, it is reference in java client API, so we cannot move it without breaking binary backward compatibility.

Maybe we can clean it up for 1.0 :thinking:

XN137 commented on model/src/main/java/org/projectnessie/api/v2/params/GetReferenceParams.java:
```diff
@@ -0,0 +1,98 @@
+
+  public GetReferenceParams() {}
+
+  @Builder.Constructor
```
in other places we use `@org.immutables.builder.Builder.Constructor`

also makes me wonder (pardon my ignorance): why are we implementing tostring/equals ourselves?
was it not possible to use immutables because we need the openapi annotations?

dimas-b replied:
copy-paste in this case. I'll see if we can use the generated methods.

dimas-b replied:
Added imports for that annotation

dimas-b replied:
I believe the hand-written classes are required for JaxRs... While the spec is not very clear on that when method parameters are concerned, I tend to think it implies having public constructors in parameter classes too.

dimas-b replied:
As for tostring/equals I do not think we need them in param classes at all.

dimas-b replied:
I've removed equals/hash code in 470df9f9138415f68a26ea2544b7c4ea861eadea

XN137 reviewed (commented):
a lot of thought and work went into this :tada: 

i've looked at a subset of the 3rd commit only here (mostly at code-surface-level) , maybe i'll get back to some more later today

i had looked at the v2 api issue but some conclusions are not super obvious since i am not familiar with all shortcomings + design considerations, snazy has probably a lot more insight from a design perspective.

i think a readme on the api history/maintenance plan would be helpful in that regard (see specific comment)

dimas-b commented on clients/client/src/main/java/org/projectnessie/client/builder/ClientSideCreateNamespace.java:
```diff
@@ -0,0 +1,92 @@
+
+      contentMap = api.getContent().reference(branch).key(key).get();
+    } catch (NessieNotFoundException | NessieConflictException e) {
+      throw new NessieReferenceNotFoundException(e.getMessage(), e);
```
I was trying to maintain old user-visible behaviour... mostly based on existing tests. I'll do another pass over this code.

dimas-b commented on clients/client/src/main/java/org/projectnessie/client/builder/ClientSideCreateNamespace.java:
```diff
@@ -0,0 +1,92 @@
+import org.projectnessie.model.Operation;
+
+/** Supports previous "create namespace" functionality of the java client over Nessie API v2. */
+public final class ClientSideCreateNamespace extends BaseCreateNamespaceBuilder {
```
It not specific to v2, could be used with v1 or another other API version that has "get entries" and "commit".

I put it in the `builder` package because it implemented the `CreateNamespaceBuilder` interface. However, the logic in this class is indeed different from other builder...

How does `...client.helper.ClientSideCreateNamespace` sound?... or `client.util`?.. Any other package name suggestions?

dimas-b replied:
Moved to `client.util.v2api` and added javadoc.

dimas-b commented on clients/client/src/main/java/org/projectnessie/client/http/ApiHttpRequest.java:
```diff
@@ -0,0 +1,64 @@
+
+import java.util.function.Supplier;
+
+public class ApiHttpRequest<E1 extends Throwable, E2 extends Throwable> {
```
sure

dimas-b replied:
javadoc added

dimas-b commented on versioned/tests/src/main/resources/META-INF/services/org.assertj.core.configuration.Configuration:
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (C) $YEAR Dremio
+# Copyright (C) 2022 Dremio
```
automatic :)

dimas-b commented on servers/quarkus-server/src/test/java/org/projectnessie/server/TestAuthorizationRules.java:
```diff
@@ -203,13 +205,15 @@ void testReadEntityAllowed() throws Exception {
   @Test
   @TestSecurity(user = "admin_user")
   void testRefLogAllowed() throws Exception {
+    Assumptions.assumeFalse(api() instanceof NessieApiV2); // Reflog is not supported in API v2
```
ATM `NessieApiV2` extends `NessieApiV1`. This extension is needed to be able to run tests against both v1 and v2 impl using the same test code.

dimas-b replied:
I've change to annotation based filtering in the latest commit in this PR. I'll probably remove this commit and rebase on top of #5436 when it is merged - just wanted to move the test infra changes into a smaller standalone PR.

snazy replied:
Nit-pick: `assumeThat(api()).isInstanceOf(NessieApiV1.class)`

dimas-b commented on servers/jax-rs-tests/src/test/java/org/projectnessie/jaxrs/TestJerseyRestNaiveClientInMemory.java:
```diff
@@ -67,8 +66,8 @@ public NessieClientBuilder<?> configure(NessieClientBuilder<?> builder) {
 
-  @AfterEach
-  void ensureHeadersProcessed() {
+  @AfterAll
```
It does break the original intention of this check, but with API v2 some test cases are suppressed by assumptions, but `@AfterEach` still runs for them and would fail. I'll probably inline the calls to `ensureHeadersProcessed`. Hope it will be clearer.

dimas-b replied:
I've changed to annotation based filtering in the latest commit in this PR. I'll probably remove this commit and rebase on top of #5436 when it is merged - just wanted to move the test infra changes into a smaller standalone PR.

dimas-b commented on servers/rest-services/src/main/java/org/projectnessie/services/rest/RestV2TreeResource.java:
```diff
@@ -0,0 +1,245 @@
+      throws NessieNotFoundException, NessieConflictException {
+    Preconditions.checkArgument(
+        transplant.keepIndividualCommits() == null,
+        "The keepIndividualCommits parameter is not supported in API v2");
```
good point.

dimas-b replied:
I reworked the first (main) commit in this PR after rebasing on top of #5435. Now setting `keepIndividualCommits` is not even possible in API v2.

dimas-b commented on model/src/main/resources/META-INF/openapi.yaml:
```diff
@@ -28,17 +28,194 @@ info:
+        
+        The 'name@hash' form always refers to the exact commit on a specific branch or tag. This is the most complete 
+        form of a reference. Other forms omit some of the details and require those gaps to be filled by the server in
+        runtime. Although these forms may be convenient to a human being, they may resolve differently at different
```
fixed

dimas-b commented on model/src/main/resources/META-INF/openapi.yaml:
```diff
@@ -28,17 +28,194 @@ info:
+        - name - identifies the HEAD commit of a branch or tag
+        - name@hash - identifies the 'hash' commit on a branch or tag
+        
+        If both 'name' and 'hash' are given, 'hash' must be the current HEAD of the branch or tag.
```
reworded

dimas-b commented on servers/jax-rs-testextension/src/main/java/org/projectnessie/jaxrs/ext/NessieJaxRsExtension.java:
```diff
@@ -173,7 +175,7 @@ public Object resolveParameter(
   @Override
   protected URI getBaseUri(ExtensionContext extensionContext) {
-    return getEnv(extensionContext).jerseyTest.target().getUri();
+    return getEnv(extensionContext).jerseyTest.target().getUri().resolve("v1");
```
reverted

dimas-b commented on clients/client/src/main/java/org/projectnessie/client/builder/ClientSideCreateNamespace.java:
```diff
@@ -0,0 +1,92 @@
+import org.projectnessie.model.Namespace;
+import org.projectnessie.model.Operation;
+
+/** Supports previous "create namespace" functionality of the java client over Nessie API v2. */
```
added javadoc

dimas-b commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRestEntries.java:
```diff
@@ -275,6 +275,8 @@ public void filterEntriesByNamespace(ReferenceMode refMode)
   @EnumSource(ReferenceMode.class)
   public void filterEntriesByNamespaceAndPrefixDepth(ReferenceMode refMode)
       throws BaseNessieClientServerException {
+    assumeApiV1();
```
I've changed to annotation based filtering in the latest commit in this PR. I'll probably remove this commit and rebase on top of #5436 when it is merged - just wanted to move the test infra changes into a smaller standalone PR.

dimas-b commented:
I've squashed some of the review-based changes into the main (first) commit because the affected code got many changes from related PRs and the interplay of changes was becoming awkward. No functional changes though.

Will do another rebase after #5419 and #5435 are merged.

snazy commented on servers/rest-services/src/main/java/org/projectnessie/services/rest/RestV2TreeResource.java:
```diff
@@ -0,0 +1,238 @@
+  }
+
+  @Override
+  public ContentResponse getContent(ContentKey key, String ref) throws NessieNotFoundException {
```
Looks like this method has no test coverage

dimas-b replied:
added `AbstractV2ResteasyTest` to cover this

snazy commented on servers/rest-services/src/main/java/org/projectnessie/services/rest/RestV2ConfigResource.java:
```diff
@@ -0,0 +1,37 @@
+@RequestScoped
+@Path("v2beta/config")
+public class RestV2ConfigResource extends RestConfigResource {
+
```
Would be nicer IMO to not extend the V1 REST class but implement the HttpConfigApi interface from the v2 package.

dimas-b replied:
reworked

snazy reviewed (commented):
Overall LGTM, it's a huge chunk of code.

The enhanced reference syntax needs low-level unit tests and likely some tests using the new reference variaants via the API and REST.

REST/Swagger docs need quite some love as well, but that's likely for a follow-up.

dimas-b commented on versioned/tests/src/main/resources/META-INF/services/org.assertj.core.configuration.Configuration:
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (C) $YEAR Dremio
+# Copyright (C) 2022 Dremio
```
Auto-correction by IntelliJ... Shall I set it to 2021?

snazy replied:
Oh, just saw _that_ there's a diff. 2022 is probably fine.

dimas-b commented on model/src/main/java/org/projectnessie/model/Validation.java:
```diff
@@ -35,17 +35,25 @@ public final class Validation {
+      "Reference name must "
+          + REF_RULE
+          + ", optionally followed "
+          + "by ~ and a commit hash, which must "
```
Yes, I think we wanted to use `~` I guess forgot to update the code. 

IMHO, `~` is preferable because it is not a special char in URLs. I will change to `~` unless there are objections :) WDYT?

snazy replied:
Either WFM

dimas-b replied:
I just released that python CLI already used the `main@11223344` syntax... I'll keep the `@` separator for now. We can change is later across all tools.

dimas-b commented on model/src/main/java/org/projectnessie/api/v2/params/Transplant.java:
```diff
@@ -0,0 +1,78 @@
+@Value.Immutable
+@JsonSerialize(as = ImmutableTransplant.class)
+@JsonDeserialize(as = ImmutableTransplant.class)
+public interface Transplant extends BaseMergeTransplant {
```
My thinking is that we do not really want end-users coding against `param` classes to reduce chances of compatibility issues arising during API evolution :thinking:

snazy replied:
Fair point

dimas-b commented on model/src/main/java/org/projectnessie/api/v2/params/Merge.java:
```diff
@@ -0,0 +1,81 @@
+@Value.Immutable
+@JsonSerialize(as = ImmutableMerge.class)
+@JsonDeserialize(as = ImmutableMerge.class)
+public interface Merge extends BaseMergeTransplant {
```
let's discuss under `Transplant` above.

dimas-b commented on model/src/main/java/org/projectnessie/api/v2/TreeApi.java:
```diff
@@ -0,0 +1,283 @@
+      @Valid Transplant transplant)
+      throws NessieNotFoundException, NessieConflictException;
+
+  /** merge mergeRef onto ref. */
```
fixed

dimas-b commented on model/src/main/java/org/projectnessie/api/v2/http/HttpTreeApi.java:
```diff
@@ -0,0 +1,637 @@
+@Consumes(value = MediaType.APPLICATION_JSON)
+@Path("trees")
+@Tag(name = "v2")
+public interface HttpTreeApi extends TreeApi {
```
moved to java interface constants and cleaned up the descriptions somewhat.

dimas-b commented on model/src/main/java/org/projectnessie/api/v2/params/DiffParams.java:
```diff
@@ -0,0 +1,58 @@
+import org.immutables.builder.Builder.Constructor;
+import org.projectnessie.model.Validation;
+
+public class DiffParams {
```
added

dimas-b commented on model/src/main/java/org/projectnessie/api/v2/params/GetReferenceParams.java:
```diff
@@ -0,0 +1,71 @@
+
+public class GetReferenceParams {
+
+  @Parameter(ref = "refPathParameter")
```
added

dimas-b commented on model/src/main/java/org/projectnessie/model/Util.java:
```diff
@@ -61,6 +62,51 @@ public static String toPathString(List<String> elements) {
         .collect(Collectors.joining("."));
   }
 
+  public static String toPathStringRef(String name, String hash) {
```
added

dimas-b commented on model/src/main/java/org/projectnessie/api/v2/TreeApi.java:
```diff
@@ -0,0 +1,283 @@
+          @Pattern(
+              regexp = Validation.REF_NAME_PATH_REGEX,
+              message = Validation.REF_NAME_PATH_MESSAGE)
+          String refName,
```
Good catch! I assumed existing tests covered that, but apparently `Detached.of(hash).getName()` is not `null` but `DETACHED` so the name-less cases are not covered by existing tests :thinking:

dimas-b replied:
should be fixed now

snazy commented on clients/client/src/main/java/org/projectnessie/client/http/v2api/HttpCommitMultipleOperations.java:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  @Override
+  public Branch commit() throws NessieNotFoundException, NessieConflictException {
```
Since we're changing the API anyway, WDYT about wrapping these "bare" types into "per operation return types", e.g. some `CommitResult` type returned by this method? Similar to the wrapper types in the OpenAPI spec. (In a follow-up)

dimas-b replied:
So far all java client changes should be backward compatible, so older clients can switch to API v2 by just declaring `NessieApiV2.class` when they create the Nessie Client.

We could certainly add a _new_ commit variant to return the more complete result, e.g. `commitAdvanced()`, WDYT?

snazy replied:
Yea - just raised this. Can you open a follow-up issue?

dimas-b replied:
#5487

snazy commented on clients/client/src/main/java/org/projectnessie/client/http/ApiHttpRequest.java:
```diff
@@ -0,0 +1,78 @@
+ * @param <E1> the first API-level exception that should be unwrapped
+ * @param <E2> the second API-level exception that should be unwrapped
+ */
+public class ApiHttpRequest<E1 extends Throwable, E2 extends Throwable> {
```
Does it need to be `public`? (Isn't package-private good enough? Maybe with a wrapping interface - but w/o over-engineering - so only, if it's a simple thing to do)

dimas-b replied:
will do another follow-up commit for this

dimas-b replied:
I extracted an interface for now - see the latest review commit.

It might be nicer just to fold the exception unwrapping logic into `HttpRequest` itself, but I did not want to touch a lot of v1 code with minor refactoring due to type parameter changes. We could do that in a follow-up PR. WDYT?

snazy commented on servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRestMergeTransplant.java:
```diff
@@ -50,6 +56,12 @@
 public abstract class AbstractRestMergeTransplant extends AbstractRestInvalid {
 
+  static {
+    ConfigurationProvider.CONFIGURATION_PROVIDER
```
Can you move it to a file similar to `versioned/tests/src/main/resources/META-INF/services/org.assertj.core.configuration.Configuration`

dimas-b replied:
I removed this code block completely. It is not needed under the new versioned test extensions.

snazy commented on model/build.gradle.kts:
```diff
@@ -51,7 +51,7 @@ smallryeOpenApi {
   operationIdStrategy.set("METHOD")
   scanPackages.set(
-    listOf("org.projectnessie.api", "org.projectnessie.api.http", "org.projectnessie.model")
+    listOf("org.projectnessie.api.v2", "org.projectnessie.api.v2.http", "org.projectnessie.model")
```
Q: Don't we need v1 here as well?

snazy replied:
maybe publish v1 + v2beta now and remove v1 later

dimas-b replied:
yes, makes sense... will do.

This will probably touch the "main" commit with some changes in API classes to re-review... just FYI :)

On the positive side, this should help to clean-up the Swagger page in Quarkus dev UI.

dimas-b replied:
switched to `v2beta` URLs

dimas-b replied:
Using `v2-beta` OpenAPI tags instead now.

snazy reviewed (commented):
Thanks! Looks much better now!
Just a few minor things and Qs and (optional) nits.

dimas-b commented:
rebased... addressing remaining comments :hourglass:

dimas-b commented:
Squashed. Rebased.

Changed v2 OpenAPI tag to `v2-beta` but went back to the simple `/v2` URL paths.

Combined v1 and v2 in the same OpenAPI yaml. This also removes the mess in the Quarkus Swagger page.

snazy commented on model/src/main/java/org/projectnessie/api/v2/params/EntriesParams.java:
```diff
@@ -0,0 +1,71 @@
+ * <p>For easier usage of this class, there is {@link EntriesParams#builder()}, which allows
+ * configuring/setting the different parameters.
+ */
+public class EntriesParams extends AbstractParams<EntriesParams> {
```
Note: you can also add start+end keys for this one along with the DiffParams follow-up.

dimas-b replied:
#5591

snazy commented on model/src/main/java/org/projectnessie/api/v2/params/ReferencesParams.java:
```diff
@@ -0,0 +1,104 @@
+ * <p>For easier usage of this class, there is {@link ReferencesParams#builder()}, which allows
+ * configuring/setting the different parameters.
+ */
+public class ReferencesParams extends AbstractParams<ReferencesParams> {
```
This one can also include start+end key equivalents.

dimas-b replied:
#5591

snazy commented on model/src/main/java/org/projectnessie/api/v2/http/HttpTreeApi.java:
```diff
@@ -0,0 +1,723 @@
+
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  @Path("{ref:" + REF_NAME_PATH_ELEMENT_REGEX + "}/contents")
```
The new storage stuff could return contents using paging, and have start+end key constraints. Can you support for paging for `getSeveralContents` + `getMultipleContents`, too?`

dimas-b replied:
Certainly. That is not directly related to the `ref` param, though, is it?

dimas-b replied:
#5591

snazy commented on gc/gc-tool/src/test/java/org/projectnessie/gc/tool/TestCLI.java:
```diff
@@ -64,6 +66,7 @@
 @NessieDbAdapterName(InmemoryDatabaseAdapterFactory.NAME)
 @NessieExternalDatabase(InmemoryTestConnectionProviderSource.class)
 @TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+@NessieApiVersions(versions = NessieApiVersion.V1)
```
Can you open a follow-up to let GC use V2?

dimas-b replied:
#5422

snazy commented on model/src/main/resources/META-INF/openapi.yaml:
```diff
@@ -28,17 +28,197 @@ info:
     url: http://www.apache.org/licenses/LICENSE-2.0.html
 
+tags:
+  - name: v1
```
Ahhh!
![image](https://user-images.githubusercontent.com/957468/204901610-6bf6df29-2944-4347-bd3f-df7bec6eeb6f.png)

snazy reviewed (commented):
Looks pretty solid and veeeeery close to get merged. Just a few comments, mostly questions though.

Maybe have someone who's not familiar with the changes do a proof-read the OpenAPI docs - in a follow-up.

dimas-b commented on model/src/test/java/org/projectnessie/model/TestReference.java:
```diff
@@ -0,0 +1,63 @@
+  })
+  void fromPathString(String pathParameter, String expectedName, String expectedHash) {
+    for (ReferenceType type : ReferenceType.values()) {
+      assertThat(Reference.fromPathString(pathParameter, type))
```
fixed

snazy commented on clients/client/src/main/java/org/projectnessie/client/http/v2api/HttpAssignBranch.java:
```diff
@@ -0,0 +1,41 @@
+        .resolveTemplate("ref", Reference.toPathString(branchName, hash))
+        .queryParam("type", Reference.ReferenceType.BRANCH.name())
+        .unwrap(NessieNotFoundException.class, NessieConflictException.class)
+        .put(assignTo);
```
Would be nice to return the value from the REST API (same for "assign tag"), can be done via a follow-up

snazy commented on clients/client/src/main/java/org/projectnessie/client/util/v2api/ClientSideDeleteNamespace.java:
```diff
@@ -0,0 +1,99 @@
+          .hash(hashOnRef)
+          .commitMeta(CommitMeta.fromMessage("delete namespace " + key))
+          .operation(Operation.Delete.of(key))
+          .commit();
```
Would be nice to return the result of the commit() (follow-up)

dimas-b replied:
That would change the java client method signature and break existing clients. We can add a new method to `UpdateNamespaceBuilder` for that, though.

dimas-b replied:
Maybe under #5487 ?

snazy commented on clients/client/src/main/java/org/projectnessie/client/util/v2api/ClientSideDeleteNamespace.java:
```diff
@@ -0,0 +1,99 @@
+    Optional<Object> existingNamespace =
+        Optional.ofNullable(contentMap.get(key)).flatMap(c -> c.unwrap(Namespace.class));
+
+    if (!existingNamespace.isPresent()) {
```
Nit: can be simplified with `.orElseThrow(() -> new ...Exception...`

dimas-b replied:
#5594

snazy commented on clients/client/src/main/java/org/projectnessie/client/util/v2api/ClientSideDeleteNamespace.java:
```diff
@@ -0,0 +1,99 @@
+      throw new NessieReferenceNotFoundException(e.getMessage(), e);
+    }
+
+    if (entry.isPresent()) {
```
Nit: can also be simplified with `.orElseThrow()`

dimas-b replied:
#5594

snazy commented on clients/client/src/main/java/org/projectnessie/client/util/v2api/ClientSideGetNamespace.java:
```diff
@@ -0,0 +1,62 @@
+
+    Optional<Namespace> result =
+        Optional.ofNullable(contentMap.get(key)).flatMap(c -> c.unwrap(Namespace.class));
+    if (!result.isPresent()) {
```
Nit: `.orElseThrow()` + inline the `return`

dimas-b replied:
#5594

snazy commented on clients/client/src/main/java/org/projectnessie/client/util/v2api/ClientSideUpdateNamespace.java:
```diff
@@ -0,0 +1,70 @@
+          .hash(hashOnRef)
+          .commitMeta(CommitMeta.fromMessage("update namespace " + key))
+          .operation(Operation.Put.of(key, builder.build()))
+          .commit();
```
Would be nice to return the result of the commit call

dimas-b replied:
That would change the java client method signature and break existing clients. We can add a new method to `UpdateNamespaceBuilder` for that, though.

snazy commented on clients/client/src/main/java/org/projectnessie/client/util/v2api/ClientSideCreateNamespace.java:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  @Override
+  public Namespace create()
```
Would be nice to get the `Branch` returned from `commit()` as a composite result (follow-up)

dimas-b replied:
#5487

snazy reviewed (approved):
Added a couple more comments for follow-ups and a few nits, nothing that blocks this PR.
Feel free to address the nits and merge w/ the dismissed approval.

dimas-b commented:
Thanks for the reviews @snazy and @XN137 !

I'll address remaining comments in follow-up PRs.
